### DISCUSSION TITLE
Custom channel layout

### DISCUSSION THREAD

**Discussion by adampavlinek** [SAND-Lab/MEA-NAP#41] [Q&A] [answered]
How can we input a custom channel layout e.g. for an Axion16 (for 24-well plates)?

**Comment by Timothysit** [comment, marked as answer]
Hi, to input a custom channel layout, first, you need to go to the tab under "spike detection" and change the layout to "Custom". Then, go to the script AnalysisPipeline/Functions/getCoordsFromLayout.m, and edit the channel numbering and coordinates after line 149 (you can use the earlier lines for "Axion64" as a reference to how the channel ordering and layout numbers work). When you start running the pipeline, it should also print out the channel layout and numbering for you to double check it is what you wanted. Let me know if you have any issues! 
<img width="638" alt="image" src="https://github.com/SAND-Lab/MEA-NAP/assets/11500169/e9a83b5c-ff4c-468e-8515-92c7da165ee3">

**Reply by adampavlinek** [reply]
Hi thanks so much! So I modified the code as follows:
elseif strcmp(channelLayout, 'Custom')
    
        channels = [11, 12, 13, 14, ... 
                 21, 22, 23, 24, ...
                 31, 32, 33, 34, ...
                 41, 42, 43, 44];
    coords = zeros(length(channels), 2);
    coords(:, 2) = repmat(linspace(0, 1, 4), 1, 4);
    coords(:, 1) = repelem(linspace(0, 1, 4), 1, 4);

end 

And the printed channel layout looks OK.
But I'm getting an error so I guess I'm not doing something right:
MAJOR WARNING: the provided list of channels has a different length than the number of channels in the data, Please check that the data has been processed correctly. For now I will just run spike detection up to the number ofavailable channels, assuming that the ordering of channels is still okay 
Index exceeds the number of array elements. Index must not exceed 1.

Error in batchDetectSpikes (line 236)
                channels = channels(1:numChannelsInData);

Error in MEApipeline (line 194)
        batchDetectSpikes(rawData, savePath, option, ExpName, Params);

**Reply by Timothysit** [reply]
Hi, may I ask what is the shape of your `.dat` and `.channels` variable in your converted raw data mat files? Currently our code assumes that `.dat` has shape (numTimeSamples, numChannels), and `.channels` has shape (numChannels, 1).

**Reply by adampavlinek** [reply]
Yes the .dat variable matches that, it has the shape (3746875, 16) where 16 is the number of channels. But the .channels variable is (1,1) so it doesn't have the right number of channels.

**Reply by Timothysit** [reply]
Did you follow the steps in the documentation to convert your Axion files?

**Reply by adampavlinek** [reply]
Yes, I followed the steps for conversion for .raw from Axion Maestro files.

**Reply by Timothysit** [reply]
Okay I think I found the solution! The original code we had for file conversion did not take account of the 24-well plate / 16 channel configuration. I have added that now, please convert your files again and let me know if that works.

**Reply by adampavlinek** [reply]
Hi it appears that file conversion is not working at all now, file conversion runs without any errors but nothing happens and no converted files are generated.

**Reply by Timothysit** [reply]
Are you using the GUI to do the file conversion?

**Reply by adampavlinek** [reply]
Yes, with the Axion.raw option.

**Reply by Timothysit** [reply]
ah that's strange, I just tested it on some axion .raw files (albeit with 60 electrodes) and it ran okay. Perhaps try removing the existing .mat files from the folder with the .raw files?

**Reply by adampavlinek** [reply]
Ok thank you it works! I think I was just testing it with a different file that was too large.
